Objects/dictobject.c

```diff
@@ -1278,41 +1424,19 @@ insertdict(PyInterpreterState *interp, PyDictObject *mp,
         /* Insert into new slot. */
         mp->ma_keys->dk_version = 0;
         assert(old_value == NULL);
-        if (mp->ma_keys->dk_usable <= 0) {
-            /* Need to resize. */
-            if (insertion_resize(interp, mp, 1) < 0)
-                goto Fail;
-        }
-
-        Py_ssize_t hashpos = find_empty_slot(mp->ma_keys, hash);
-        dictkeys_set_index(mp->ma_keys, hashpos, mp->ma_keys->dk_nentries);
 
-        if (DK_IS_UNICODE(mp->ma_keys)) {
-            PyDictUnicodeEntry *ep;
-            ep = &DK_UNICODE_ENTRIES(mp->ma_keys)[mp->ma_keys->dk_nentries];
-            ep->me_key = key;
-            if (mp->ma_values) {
-                Py_ssize_t index = mp->ma_keys->dk_nentries;
-                _PyDictValues_AddToInsertionOrder(mp->ma_values, index);
-                assert (mp->ma_values->values[index] == NULL);
-                mp->ma_values->values[index] = value;
-            }
-            else {
-                ep->me_value = value;
+        int unicode = DK_IS_UNICODE(mp->ma_keys);
+        if (!unicode || !_PyDict_HasSplitTable(mp)) {
```

Comment:
This can simply check that the dict does not have a split table. The `!unicode` check is redundant.

Additionally, I'd suggest moving the `DK_IS_UNICODE(mp->ma_keys)` check inside the body of `insert_combined_dict` and getting rid of the `unicode` argument. I was a bit confused as to whether that argument indicated a unicode-only table or a new unicode key.


```suggestion
        if (!_PyDict_HasSplitTable(mp)) {
```